KEP-4753: Expose direct ownerReferences via downward API (#4754)
kubernetes/enhancements: keps/sig-node/4753-expose-ownerref-downard-api/README.md

  - [ ] (R) Minimum Two Week Window for GA e2e tests to prove flake free
- [ ] (R) Graduation criteria is in place
  - [ ] (R) [all GA Endpoints](https://github.com/kubernetes/community/pull/1806) must be hit by [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md) 
- [ ] (R) Production readiness review completed
- [ ] (R) Production readiness review approved
- [ ] "Implementation History" section is up-to-date for milestone
- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes

<!--
**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
-->

[kubernetes.io]: https://kubernetes.io/
[kubernetes/enhancements]: https://git.k8s.io/enhancements
[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
[kubernetes/website]: https://git.k8s.io/website

## Summary

> Today when a pod wants to pass its 'onwerReferences' to a new object it manage (a ConfigMap, a Secret, etc), it needs to do it call the API server to get it's own ownerReferences and then pass it to the new object. 

Reviewer: ```suggestion
Today when a pod wants to pass its 'ownerReferences' to a new object it manage (a ConfigMap, a Secret, etc), it needs to do it call the API server to get it's own ownerReferences and then pass it to the new object. 
```

Author: done

The pod then needs GET access on all the intermediate resources which it may not have, and giving it that access is a security risk as it can access other resources it should not have access to.
This KEP proposes to expose the ownerReferences via `valueFrom` and downward API, so that the pod can directly pass its `ownerReferences` to the managed/created new object without having to call the API server.

This KEP is intentionally scoped small, and intends to keep the scope just to `ownerReferences` and not to other fields of the owner object to prevent security risks.

## Motivation

> Deployments and ReplicaSets can orphan and adopt pods dinamically. Any additional information we provide can change over time. On day 1 a pod named "foobar-deadbeef93-xyz76" is owned by replicaset "foobar-deadbeef93" which is owned by deployment "foobar", but after a node restart, the pod name could change to "foobar-deadbeef93-abc12", leaving the object created by "foobar-deadbeef93-xyz76" orphan, and triggering unwanted behavior. Enabling the pod to pass its ownerReferences to the new object it manages will prevent objects from being orphaned by pods that are owned by a higher level object.

Reviewer: ```suggestion
Deployments and ReplicaSets can orphan and adopt pods dynamically. Any additional information we provide can change over time. On day 1 a pod named "foobar-deadbeef93-xyz76" is owned by replicaset "foobar-deadbeef93" which is owned by deployment "foobar", but after a node restart, the pod name could change to "foobar-deadbeef93-abc12", leaving the object created by "foobar-deadbeef93-xyz76" orphan, and triggering unwanted behavior. Enabling the pod to pass its ownerReferences to the new object it manages will prevent objects from being orphaned by pods that are owned by a higher level object.
```

Can you add new lines to this paragraph also?

> Deployments and ReplicaSets can orphan and adopt pods dinamically. Any additional information we provide can change over time. On day 1 a pod named "foobar-deadbeef93-xyz76" is owned by replicaset "foobar-deadbeef93" which is owned by deployment "foobar", but after a node restart, the pod name could change to "foobar-deadbeef93-abc12", leaving the object created by "foobar-deadbeef93-xyz76" orphan, and triggering unwanted behavior. Enabling the pod to pass its ownerReferences to the new object it manages will prevent objects from being orphaned by pods that are owned by a higher level object.

Reviewer 2: I'm afraid this motivation doesn't make much sense to me.  I think you've got some context in your head that is not in the text.

Why does a pod need to know it's own ownerRefs ?  You seem to be implying that the pod will create some other object (e.g. a ConfigMap) and rather than parent that object itself, it wants to create the ConfigMap and set its ownerRef to the pod's ownerRef?  Why?

It seems plausible to me that a pod would be able to know its own ownerRefs, but that use-case eludes me.

Help me understand what a pod will actually DO with this information?

Author: how about now?

Author: ping @thockin

### Goals

* Gain access to pod `ownerReferences` on pods through volume mounts.
* Gain access to pod `ownerReferences` on pods through environmental variables.

### Non-Goals

- Not to expose additional pod info outside of `metadata.ownerReferences`.

## Proposal

The initial design includes:

* Being able to pass pod `ownerReferences` to volumes with fieldPath of `metadata.ownerReferences`.
* Being able to pass pod `ownerReferences` to environmental variables with fieldPath of `metadata.ownerReferences`.

### User Stories (Optional)

#### Story 1
xref https://github.com/kubernetes-sigs/node-feature-discovery/issues/1752

> CustomResource object created by a pod owned by a DaemonSet is garbage collected when the pod is deleted by unexpected reasons. This triggers unwanted behavior in the system.

Reviewer 2: Ahh, the DaemonSet part helps clarify a little.  Can you make this even more concrete?

If the CustomResource object inherits the ownerReferences from the pod, it will be garbage collected only when the DaemonSet is deleted. This will add extra reliability to the system by projecting the ownerReferences from the Deployment/DaemonSet/StatefulSet to the CustomResource object.

### Notes/Constraints/Caveats (Optional)

<!--
What are the caveats to the proposal?
What are some important details that didn't come across above?
Go in to as much detail as necessary here.
This might be a good place to talk about core concepts and how they relate.
-->

### Risks and Mitigations

- Security Risk: Some environments may not want to expose the ownerReferences of a pod to the pod itself.
  - Mitigation: This feature could be added as a feature gate, disabled by default.

## Design Details

The intention behind this design is not to add extra structs or fields
but to extend the existing `downward API` to include the `metadata.ownerReferences` field.

> ### Implementation Details

Reviewer 2: Can you please say more about how this will present both as a file and in an env?

Can a user request a specific index of ownerRefs or does it always get all of them?  

What does the value of the env var actually look like?  OwnerRef is a struct with several fields, are we saying it will be JSON-formatted?  Or some other bespoke format?  What if I have 2 or 5 or 10 owners?  What if I have 0 owners.

Same question, but for file.  If I `cat` that file - what do I see?

Author: How about now?

> * Will use pre-existing information from the Kubelet and VolumeHost structs

Reviewer: Please make sure to use the feature gate for adding new fields to the set and to convert the downward API.

- `<package>`: `<date>` - `<test coverage>`

##### Integration tests

<!--
Integration tests are contained in k8s.io/kubernetes/test/integration.
Integration tests allow control of the configuration parameters used to start the binaries under test.
This is different from e2e tests which do not allow configuration of parameters.
Doing this allows testing non-default options and multiple different and potentially conflicting command line options.
-->

<!--
This question should be filled when targeting a release.
For Alpha, describe what tests will be added to ensure proper quality of the enhancement.

For Beta and GA, add links to added tests together with links to k8s-triage for those tests:
https://storage.googleapis.com/k8s-triage/index.html
-->

- <test>: <link to test coverage>

##### e2e tests

<!--
This question should be filled when targeting a release.
For Alpha, describe what tests will be added to ensure proper quality of the enhancement.

For Beta and GA, add links to added tests together with links to k8s-triage for those tests:
https://storage.googleapis.com/k8s-triage/index.html

We expect no non-infra related flakes in the last month as a GA graduation criteria.
-->

> - <test>: <link to test coverage>

Reviewer: I'd list out the tests you are planning on adding.

> ### Graduation Criteria

Reviewer: Given that this is not a API change, you may be able to skip alpha stage. WDYT @thockin?

Reviewer 2: It is an API change - a value that used to be rejected is now allowed, so in the event of a rollback, we need to know it is still handled.  Ergo "alpha" is required.

  - [ ] (R) Minimum Two Week Window for GA e2e tests to prove flake free
- [ ] (R) Graduation criteria is in place
  - [ ] (R) [all GA Endpoints](https://github.com/kubernetes/community/pull/1806) must be hit by [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md) 
- [ ] (R) Production readiness review completed
- [ ] (R) Production readiness review approved
- [ ] "Implementation History" section is up-to-date for milestone
- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes

<!--
**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
-->

[kubernetes.io]: https://kubernetes.io/
[kubernetes/enhancements]: https://git.k8s.io/enhancements
[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
[kubernetes/website]: https://git.k8s.io/website

## Summary

> Today when a pod wants to pass its `ownerReferences` to a new object it manage (a ConfigMap, a Secret, etc), it needs to do it call the API server to get it's own ownerReferences and then pass it to the new object. 

Reviewer 3: ```suggestion
Today, when a pod wants to pass its `ownerReferences` to a new object it manages (a ConfigMap, a Secret, etc), the pod needs it to call the API server to get its own ownerReferences and then pass it to the new object. 
```

> The pod then needs GET access on all the intermediate resources which it may not have, and giving it that access is a security risk as it can access other resources it should not have access to.

Author: What this line tries to say is that TODAY, for a pod to know its owner, it needs to do a GET call to the API Server.  getting all the info that perhaps we don't want the pod to know. And only GET bc today the pod can get the namespace and name via downwardAPI, so with those 2 fields you can do a targeted GET.

> The pod then needs GET access on all the intermediate resources which it may not have, and giving it that access is a security risk as it can access other resources it should not have access to.

Reviewer 2: Asking again - is the goal of this KEP to expose the pod's DIRECT owner or the "true" owner?  Elsewhere it was suggested that the direct owner is sufficient, but this summary suggests it actually wants the "true" owner (which is hard to spec).

Exposing a `fieldRef` can ONLY give you the direct owner -- read from fields in the pod.  It cannot give you the pod's owner's owner (eg Pod->ReplicaSet->Deployment).

Is this OK?

Otherwise the only workable answer is to make the pod ask the apiserver itself, either in the main app or in a sidecar, and to give the pod's SA enough permission to get the info it needs.  We CANNOT give the power to read arbitrary resources to kubelet.

> The pod then needs GET access on all the intermediate resources which it may not have, and giving it that access is a security risk as it can access other resources it should not have access to.

Author: The goal is to avoid an API call (GET) but instead have the info at hand if needed to be passed down. not the true owner but the pod DIRECT owner
so:
UNKNOWN - > Daemonset -> Pod -> Object

> The pod then needs GET access on all the intermediate resources which it may not have, and giving it that access is a security risk as it can access other resources it should not have access to.

Author: it now reads more detailed

This KEP proposes to expose the ownerReferences via `valueFrom` and downward API, so that the pod can directly pass its `ownerReferences` to the managed/created new object without having to call the API server.

This KEP is intentionally scoped small, and intends to keep the scope just to `ownerReferences` and not to other fields of the owner object to prevent security risks.

## Motivation

> DaemonSets and ReplicaSets (via Deployments) can dynamically orphan and adopt pods. Consider a scenario where, on day 1, a pod named foobar-deadbeef93-xyz76 is owned by the ReplicaSet foobar-deadbeef93, which in turn is managed by the Deployment foobar. After a node restart, the pod might be replaced by foobar-deadbeef93-abc12, orphaning the original pod foobar-deadbeef93-xyz76 and potentially causing unwanted behavior. This issue arises when pods create resources like ConfigMaps or CustomResources and these resources are not correctly reassigned to the new pod, leading to orphaned objects. Ensuring that pods can pass their ownerReferences to new objects they manage would prevent this issue, as the ownership hierarchy would be preserved even when pods are recreated.

Reviewer 4: >  After a node restart, the pod might be replaced by foobar-deadbeef93-abc12, orphaning the original pod foobar-deadbeef93-xyz76 and potentially causing unwanted behavior

can this really happen? it this not a bug on the replicaset?

Author: You are right, and the motivation that inspired me to propose this KEP is based on a DaemonSet use case, I have updated the text

> For example, a CustomResource created by a pod managed by a DaemonSet may be unexpectedly garbage collected if the pod is deleted. This can disrupt system behavior, as the necessary resource is lost. By allowing the pod to inherit and pass down ownerReferences, the CustomResource would remain correctly managed by the DaemonSet, avoiding such disruptions.

Author: I have rephrased this line, does it convey the intention now?

* Being able to pass pod `ownerReferences` to environmental variables with fieldPath of `metadata.ownerReferences`.

### User Stories (Optional)

#### Story 1
xref https://github.com/kubernetes-sigs/node-feature-discovery/issues/1752

CustomResource object created by a pod owned by a DaemonSet is garbage collected when the pod is deleted by unexpected reasons. This triggers unwanted behavior in the system.
If the CustomResource object inherits the ownerReferences from the pod, it will be garbage collected only when the DaemonSet is deleted. This will add extra reliability to the system by projecting the ownerReferences from the Deployment/DaemonSet/StatefulSet to the CustomResource object.

### Notes/Constraints/Caveats (Optional)

<!--
What are the caveats to the proposal?
What are some important details that didn't come across above?
Go in to as much detail as necessary here.
This might be a good place to talk about core concepts and how they relate.
-->

### Risks and Mitigations

- Security Risk: Some environments may not want to expose the ownerReferences of a pod to the pod itself.

>   - Mitigation: This feature could be added as a feature gate, disabled by default.

Reviewer 2: This is not a mitigation to this KEP.  If there is some cluster wherein the admin DOES NOT want pods to be able to know their owners, the mitigation would be to install an admission check (webhook or Validating Admission Policy) to reject pods that try to use this.

Author: I have edited the Mitigation entry, thanks @thockin

	}
}
```
* Information will be passed to the pod via the `valueFrom` field in the volume.

```yaml
apiVersion: v1
kind: Pod
metadata:
  name: kubernetes-downwardapi-volume-example
  ownerReferences:
  - apiVersion: apps/v1
    blockOwnerDeletion: true
    controller: true
    kind: DaemonSet
    name: an-owned-pod-1722852739
    uid: 6be1683f-da9c-4f68-9440-82376231cfa6
  - apiVersion: apps/v1
    blockOwnerDeletion: true
    controller: true
    kind: Some-CRD
    name: an-owned-pod-1722852739
    uid: 6be1683f-da9c-4f68-9440-82376231cfa6
spec:
  containers:
    - name: client-container
      image: busybox
      command: ["sh", "-c"]
      args:
      - while true; do
          if [[ -e /etc/podinfo/ownerReferences ]]; then
            echo -en '\n\n'; cat /etc/podinfo/ownerReferences; fi;
          sleep 5;
        done;
      volumeMounts:
        - name: podinfo
          mountPath: /etc/podinfo
  volumes:
    - name: podinfo
      downwardAPI:
        items:
          - path: "ownerReferences"
            fieldRef:
              fieldPath: metadata.ownerReferences
```

Logs from the pod will show the ownerReferences of the pod:

```bash

> $ kubectl logs kubernetes-downwardapi-volume-example

Reviewer 2: This is under-specified.

I don't know if we should dump the WHOLE object, but it's probably fine - can you think of any reason to hide any of these?

YAML is a terrible file format for parsing - I think it would be better to spec this as something like:

"""
The file representation of this fieldref is a JSON list of objects, each of which is a serialized `k8s.io/apimachinery/pkg/apis/meta/v1.OwnerReference`.
"""

Open question:  Should we instead produce a metav1.List?  e.g.

```
{
    "kind": "OwnerReference",
    "apiVersion": "meta/v1",
    "items": [
        { .... }, { ... }
    ]
}
```

IF there was ever a meta/v2, we would need to allow people to request that - does the output need to be self-documenting?  I think probably not.

Author: > I don't know if we should dump the WHOLE object, but it's probably fine - can you think of any reason to hide any of these?

I think if we go for your JSON recommendation and only pass a list is more than ok, all we need is the ownerRef to pass it to objects that the Pod will create. 

So to your open question, I like the `metav1.List` idea.

Reviewer 2: another ping to @deads2k or @jpbetz

Reviewer 5: Thanks for the poke.  I like the idea of using `metav1.List` to represent this.  It's a little unusual for items in a `metav1.List` to not contain their own `TypeMeta` (and `OwnerReferences` doesn't have `TypeMeta`), but on balance I prefer it to leaving the file without self-identification.

- apiVersion: apps/v1
  blockOwnerDeletion: true
  controller: true
  kind: DaemonSet
  name: an-owned-pod-1722852739
  uid: 6be1683f-da9c-4f68-9440-82376231cfa6
- apiVersion: apps/v1
  blockOwnerDeletion: true
  controller: true
  kind: Some-CRD
  name: an-owned-pod-1722852739
  uid: 6be1683f-da9c-4f68-9440-82376231cfa6
```

> If `ownerReferences` is empty or not set, the file will be empty.

Reviewer 4: what happen if `ownerReferences` is updated? is it reflected in the Pod or the Pod keeps using the original value?

Author: If this were to happen, isn't the Daemonset responsibility to refresh all the owned pods? and during this update cycle the injected file would be updated.

Reviewer 6: I know very little about sig apps side of it. But is there any mutable fields on replicaset or DaemonSet that would not change the pod spec? Change of those will not be reflected in the downward API object.

Whenever we have any mutable value we lock in time, we like to add something in status to indicate which version of that value we have written down.

Maybe I misunderstood @aojea 's question.

Reviewer 6: Also behavior on Container restart would be interesting. Whether we lock the value in time of Pod creation or we re-populate on every Container restart

Reviewer 2: Env variables obviously cannot be updated on running pods, but the volume projections do auto-update.  So I would assume that if OwnerReferences changes, projections change with it.

Using env vars is a risk people take, hopefully when they know that the data they want won't actually change.

I DO NOT think we should be restarting pods when their EnvVar source changes. :)

> Is out of the scopt of this KEP to add the `ownerReferences` to the pod's environment variables.

Reviewer 4: ```suggestion
Is out of the scope of this KEP to add the `ownerReferences` to the pod's environment variables.
```

> Is out of the scope of this KEP to add the `ownerReferences` to the pod's environment variables.

Reviewer 2: Why?  It's always a smell when we do one but not the other.  This value could be specced the same as the file - a JSON list.

### Test Plan

<!--
**Note:** *Not required until targeted at a release.*
The goal is to ensure that we don't accept enhancements with inadequate testing.

All code is expected to have adequate tests (eventually with coverage
expectations). Please adhere to the [Kubernetes testing guidelines][testing-guidelines]
when drafting this test plan.

[testing-guidelines]: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
-->

[x] I/we understand the owners of the involved components may require updates to
existing tests to make this code solid enough prior to committing the changes necessary
to implement this enhancement.

##### Prerequisite testing updates

<!--
Based on reviewers feedback describe what additional tests need to be added prior
implementing this enhancement to ensure the enhancements have also solid foundations.
-->

##### Unit tests

Unit and e2e testing will be added consistent with other resources in downward API.

##### Integration tests

There are no integration tests, only e2e tests.

##### e2e tests

e2e testing will consist of the following:
- Create a Deployment and or DaemonSet

> - Verify if the pod created by the Deployment/DaemonSet has the ownerReferences of the Deployment/DaemonSet via ENV VAR

Reviewer 4: is this correct? you said before env variables are out of scope

> Is out of the scopt of this KEP to add the `ownerReferences` to the pod's environment variables.

Author: my bad fixed.

Author: fixed

- Two versions passed since introducing the functionality that deprecates the flag (to address version skew)
- Address feedback on usage/changed behavior, provided on GitHub issues
- Deprecate the flag
-->

### Upgrade / Downgrade Strategy

<!--
If applicable, how will the component be upgraded and downgraded? Make sure
this is in the test plan.

Consider the following in developing an upgrade/downgrade strategy for this
enhancement:
- What changes (in invocations, configurations, API use, etc.) is an existing
  cluster required to make on upgrade, in order to maintain previous behavior?
- What changes (in invocations, configurations, API use, etc.) is an existing
  cluster required to make on upgrade, in order to make use of the enhancement?
-->

### Version Skew Strategy

There is no issue with the version skew.

## Production Readiness Review Questionnaire

### Feature Enablement and Rollback

Simple change of a feature gate will either enable or disable this feature.

###### How can this feature be enabled / disabled in a live cluster?

- [x] Feature gate (also fill in values in `kep.yaml`)
  - Feature gate name: `DownwardAPIOwnerReferences`

>   - Components depending on the feature gate: `kubelet`

Reviewer 7: Why cannot this be a new field in kubelet configuration?
If the feature (implementation) turns out to be not appropriate, users can simply update the kubelet configuration and restart `kubelet`. However, if it is a feature gate, users don't know which components are affected. We don't have documentations elsewhere stating that this gate only affects `kubelet`. The best bet for them would be to restart all kube-xxx components because, this gate (as with others) will show up in `kube-apiserver --help` anyway when landed.

We have witnessed a lot of "abuse" of feature gates. Can we think twice about this?

Reviewer 4: feature gate is a mechanism to introduce new changes to mitigate the risks of breaking existing clusters, they are not meant to be used as flags, once they graduate they disappear and the functionality can not be disabled ... are you implying this KEP should add the functionality as opt-in or opt-out via configuration?

Reviewer 7: I think I know what feature gate was designed for ... My question is that do we want all new changes to be introduced with a feature gate? All changes could break existing clusters. I raised this concern because we are seeing that quite some new feature gates introduced lately are about trivial changes. e.g. new optimizations to the watch cache, adding new fields to the status of a resource, adding new entry to resource annotation ...

Maybe this one is not so trivial, maybe it actually warrants a new feature gate that takes several releases to become GA. My point is about the user-facing changes. For most cluster admins, feature gates are currently applied system wide, across all components. That is a HEAVY burden. I hope our team will consider this factor.

Reviewer 2: > do we want all new changes to be introduced with a feature gate?

If you think your change could break a cluster, then YES.

If you your change has any aspect of API, then YES.

> new optimizations to the watch cache

Could absolutely break a cluster!

> adding new fields to the status of a resource

Any API field we add goes through a gate so that we can safely do a rollback without breaking or dropping information.

> adding new entry to resource annotation

If this produces a value which version x-1 of Kubernetes would reject, then it needs a gate.  Always think about this case:

User upgrades cluster from x->y
User uses your new API field.
Uh oh!  Something bad happened.
User rolls cluster back to x.

Your API field is still stored in etcd, and may have some lingering external effect - what will happen?  What if the user tries to update the object - will it fail validation?  Or will the new field just disappear?

Author: we agree that this change is ok to be done via feature gates? can we mark this comment resolved?

###### Does enabling the feature change any default behavior?

No

###### Can the feature be disabled once it has been enabled (i.e. can we roll back the enablement)?

> No, the missing Downward API field will be perceived as a missing Volume preventing the pod from starting.

Reviewer 2: Actually, wouldn't it be seen as an unknown field ref?

Author: you are right.  fixed

<!--
**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
-->

[kubernetes.io]: https://kubernetes.io/
[kubernetes/enhancements]: https://git.k8s.io/enhancements
[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
[kubernetes/website]: https://git.k8s.io/website

## Summary

Today when a pod wants to pass its `ownerReferences` to a new object it manage (a ConfigMap, a Secret, etc), it needs to call the API server to GET it's own ownerReferences and then pass it to the new object. 
The pod then needs GET access on all the intermediate resources which it may not have, and giving it that access is a security risk as it can access other resources it should not have access to.
This KEP proposes to expose the ownerReferences via `valueFrom` and downward API, so that the pod can directly pass its `ownerReferences` to the managed/created new object without having to call the API server.

This KEP is intentionally scoped small, and intends to keep the scope just to `ownerReferences` and not to other fields of the owner object to prevent security risks.

## Motivation

> DaemonSets can dynamically orphan and adopt pods. Consider a scenario where, on day 1, a pod named foobar-deadbeef93-xyz76 on node XYZ creates an object (e.g a ConfigMap) associated with node XYZ, then  it would be incorrect to set the object owner to the pod foobar-deadbeef93-xyz76 instead, it should be set to damonset foobar, to prevent the object from being garbage collected if the pod cease to exist. This issue arises when pods create resources like ConfigMaps or CustomResources and these resources are not correctly reassigned to the new pod, leading to orphaned objects. Ensuring that daemonset pods can pass their ownerReferences to new objects they manage would prevent this issue, as the ownership hierarchy would be preserved even when pods are recreated.

Author: Answers to:
1- on my use case we have a Garbage collector that will remove CR's that don't have a node (Node down or gone)
2- No, for security reasons, we have the responsibilities separated, so the pod from the DaemsonSet is the only one with RBAC to create per-node CR's.
3- Yes, garbage collecting all CR's if the DaemsonSet goes away is ok as per our use case

> DaemonSets can dynamically orphan and adopt pods. Consider a scenario where, on day 1, a pod named foobar-deadbeef93-xyz76 on node XYZ creates an object (e.g a ConfigMap) associated with node XYZ, then  it would be incorrect to set the object owner to the pod foobar-deadbeef93-xyz76 instead, it should be set to damonset foobar, to prevent the object from being garbage collected if the pod cease to exist. This issue arises when pods create resources like ConfigMaps or CustomResources and these resources are not correctly reassigned to the new pod, leading to orphaned objects. Ensuring that daemonset pods can pass their ownerReferences to new objects they manage would prevent this issue, as the ownership hierarchy would be preserved even when pods are recreated.

Reviewer 2: With 3, I meant What about Deployment or what about the (hypothetical) `DaemonSetSet` which manifests DaemonSets the same way Deployment manifests ReplicaSets?  Does the ownerRef need to be the DS or the DSS ?

To be clear - I am not actually objecting to this KEP on the whole, just wondering what the NEXT feature request will look like.  Doing indirect owners is an entirely different problem.

> DaemonSets can dynamically orphan and adopt pods. Consider a scenario where, on day 1, a pod named foobar-deadbeef93-xyz76 on node XYZ creates an object (e.g a ConfigMap) associated with node XYZ, then  it would be incorrect to set the object owner to the pod foobar-deadbeef93-xyz76 instead, it should be set to damonset foobar, to prevent the object from being garbage collected if the pod cease to exist. This issue arises when pods create resources like ConfigMaps or CustomResources and these resources are not correctly reassigned to the new pod, leading to orphaned objects. Ensuring that daemonset pods can pass their ownerReferences to new objects they manage would prevent this issue, as the ownership hierarchy would be preserved even when pods are recreated.

Author: the `ownerRef` should only be the direct owner, meaning the DS, not a higher owner.

For example, a CustomResource created by a pod managed by a DaemonSet may be unexpectedly garbage collected if the pod is deleted. This can disrupt system behavior, as the CustomResource is garbageCollected. By allowing the pod to inherit and pass down ownerReferences, the CustomResource would remain correctly managed by the DaemonSet(the owner), avoiding disruptions.

> ### Goals

Reviewer 6: do we only need one level of ownerReference? Would it ever be needed to know the owner of the owner of the owner? I cannot tell from the motivation section whether one level will always be enough

Author: Well, I don't have a strong use case for that. If you have a use case maybe we can discuss it and integrate it.

Reviewer 6: one option is to just understand how it can be extended to the owner of an owner in future. To make sure we are not making it unnatural to be extended this way in future

Reviewer 2: The Summary of this KEP suggests that owner of owner is the goal.

Author: yes owner of the owner

UNKNOWN - > Daemonset/Deployment -> Pod -> Object

Author: it now reads better

### Notes/Constraints/Caveats (Optional)

<!--
What are the caveats to the proposal?
What are some important details that didn't come across above?
Go in to as much detail as necessary here.
This might be a good place to talk about core concepts and how they relate.
-->

### Risks and Mitigations

- Security Risk: Some environments may not want to expose the ownerReferences of a pod to the pod itself.
  - Mitigation: Install an admission check (webhook or Validating Admission Policy) to reject pods that try to use Downward API to access ownerReferences.

## Design Details

The intention behind this design is not to add extra structs or fields
but to extend the existing `downward API` to include the `metadata.ownerReferences` field.

### Implementation Details

* Will use pre-existing information from the Kubelet and VolumeHost structs
to extract pod metadata information. 

At `/pkg/apis/core/validation/validation.go` we will edit the following:
```go

> var validVolumeDownwardAPIFieldPathExpressions = sets.New(

Author: I have added also the EnvVar `DownwardAPIFieldPathExpressions`

    name: an-owned-pod-1722852739
    uid: 6be1683f-da9c-4f68-9440-82376231cfa6
  - apiVersion: apps/v1
    blockOwnerDeletion: true
    controller: true
    kind: Some-CRD
    name: an-owned-pod-1722852739
    uid: 6be1683f-da9c-4f68-9440-82376231cfa6
spec:
  containers:
    - name: client-container
      image: busybox
      command: ["sh", "-c"]
      args:
      - while true; do
          if [[ -e /etc/podinfo/ownerReferences ]]; then
            echo -en '\n\n'; cat /etc/podinfo/ownerReferences; fi;
          sleep 5;
        done;
      volumeMounts:
        - name: podinfo
          mountPath: /etc/podinfo
  volumes:
    - name: podinfo
      downwardAPI:
        items:
          - path: "ownerReferences"
            fieldRef:
              fieldPath: metadata.ownerReferences
```

Logs from the pod will show the ownerReferences of the pod:

The file representation of this fieldref is a JSON list of objects, each of which is a serialized `k8s.io/apimachinery/pkg/apis/meta/v1.OwnerReference`.

```bash
$ kubectl logs kubernetes-downwardapi-volume-example
{
    "kind": "OwnerReference",
    "apiVersion": "meta/v1",
    "items": [
        { .... }, { ... }
    ]
}
```

> If `ownerReferences` is empty or not set, the file will be empty.

Reviewer 2: Also, I would not say the file is empty, but the list of items is empty.

```
{
    "kind": "OwnerReference",
    "apiVersion": "meta/v1",
    "items": [ ]
}
```

  - [ ] (R) Minimum Two Week Window for GA e2e tests to prove flake free
- [ ] (R) Graduation criteria is in place
  - [ ] (R) [all GA Endpoints](https://github.com/kubernetes/community/pull/1806) must be hit by [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md) 
- [ ] (R) Production readiness review completed
- [ ] (R) Production readiness review approved
- [ ] "Implementation History" section is up-to-date for milestone
- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes

<!--
**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
-->

[kubernetes.io]: https://kubernetes.io/
[kubernetes/enhancements]: https://git.k8s.io/enhancements
[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
[kubernetes/website]: https://git.k8s.io/website

## Summary

> Today when a pod wants to pass its `ownerReferences` to a new object it manage (a ConfigMap, a Secret, etc), it needs to call the API server to GET it's own ownerReferences and then pass it to the new object. 

Reviewer 2: ```suggestion
Sometimes pods, especially those in a DaemonSet, need to create and manage resources.  When those pods set the `ownerReferences` on those objects, they don't want to use themselves as the owner.  They want to get their own `ownerReferences` and pass those on to the objects they manage.  Today, that means it needs to call the API server to GET it's own Pod, to extract `ownerReferences`. 
```

      how can an operator troubleshoot without logging into a master or worker node?
    - Mitigations: What can be done to stop the bleeding, especially for already
      running user workloads?
    - Diagnostics: What are the useful log messages and their required logging
      levels that could help debug the issue?
      Not required until feature graduated to beta.
    - Testing: Are there any tests for failure mode? If not, describe why.
-->

###### What steps should be taken if SLOs are not being met to determine the problem?

## Implementation History

<!--
Major milestones in the lifecycle of a KEP should be tracked in this section.
Major milestones might include:
- the `Summary` and `Motivation` sections being merged, signaling SIG acceptance
- the `Proposal` section being merged, signaling agreement on a proposed design
- the date implementation started
- the first Kubernetes release where an initial version of the KEP was available
- the version of Kubernetes where the KEP graduated to general availability
- when the KEP was retired or superseded
-->

## Drawbacks

Not that we can think of.

## Alternatives

> Today pods must use the API server to get their ownerReferences and pass them to the

Author: Currently, my user story is implementing this, I'll close this KEP for now as I agree with you, with a reduced scope of the KEP, it doesn't makes much sense.

/close
